[Discussion] - Contributors guide new resource vs inline (#19129)
hashicorp/terraform-provider-azurerm: contributing/topics/guide-new-resource-vs-inline.md

> # Guide: When to choose for an inline property (block) versus a new resource
> 
> Sometimes it might be a bit unclear whether it is necessary to create a new resource versus to add a new property or property block to an existing resource.
> 
> To get a bit of insight in how the decission will be made, this are some rules of thumb to decide. In case it is unclear, please contact one of the HashiCorp Maintainers.

Reviewer: ```suggestion
# Guide: When to inline new functionality (as either a block or property) versus a new resource

Sometimes when implementing new functionality it can be a bit unclear whether it is necessary to create a new resource versus to add a new property or block to an existing resource. 

To get a bit of insight in how a decision can be made, this are some rules of thumb to decide. In case it is unclear, please contact one of the HashiCorp Maintainers.

```

## Inline

> Most extra functionality will end up inline an existing resource. 
> 
> Properties to enable or disable functionality, define the resource functionality or configure details on a resource are most of the time inlined. Relations between resources with clearly separate concern (i.e. which VNet a K8s cluster will land in) are most of the time inlined.

Reviewer: ```suggestion
Most additional functionality will end up inline an existing resource. 

APIs to enable or disable functionality, define the resource functionality or configure details on a resource are most of the time inlined. Relations between resources with clearly separate concern (i.e. which VNet a K8s cluster will land in) are most of the time inlined.
```

- It does not have it's own API endpoint but uses the parent resource endpoint
- It does not cross security/team boundaries in the most common client situation
- It does not contain backwards compatibility issues

### Category 3: relations between resources
- Resources are really separate and have `1:many` and `many:1` relations with a (i.e. a relation property with a resource within a completely different Resource Provider `/subscriptions/<subscriptionId>/resourceGroups/<resourceGroup>/providers/Microsoft.Network/networkSecurityGroups/example-nsg` and `/subscriptions/<subscriptionId>/resourceGroups/<resourceGroup>/providers/Microsoft.Storage/storageAccounts/example-storage`: which `azurerm_storage_account` is used to store the logs from a resource).
- These relations are created with API calls to the original resource provider, not the connected one
- Reading the connections between the resources does not require extra permissions than previously necessary to create the resource (i.e. the Service Principal used to read the connection/relation should be also `Owner`/`Contributor` on the resource group the connection is made to)

## Separate resource

> While inlining might make a lot of sense for most properties, there are also good reasons to seperate it out. These arguments may not be conclusive, but can help steer in the right direction.

Reviewer: ```suggestion
While inlining might make a lot of sense for many APIs, there are also good reasons to separate them out. These arguments may not be conclusive, but can help steer in the right direction.
```
